MEN-4618: Package structure, README(s), and update_flow_control example

lluiscampos opened the pull request.

Commit:
MEN-4618: Package structure, README(s), and update_flow_control example

Signed-off-by: Lluis Campos <lluis.campos@northern.tech>

lluiscampos commented:
Originally I was not planning to do any kind of error handling (first commit) but then decided to add some basic handling: to print the error and exit (second commit). Do you guys think the code gets "too complicated" with two queues?

If so, I will drop the second commit. If not, I will squash into a single commit.

lluiscampos commented on examples/update_flow_control/update_flow_control.py:
```diff
@@ -0,0 +1,254 @@
+from pydbus import SystemBus  # type: ignore
+
+UPDATE_CONTROL_MAP_PAUSE_ALL = """{
+	"priority": -1,
```
Should we remove the `priority`?

kacf replied:
You can include it. This is the best place to show it off I think. But then you should explain what it does. Take a look at [this comment](https://github.com/mendersoftware/mender-docs/pull/1473#discussion_r652472821) (never mind that it was removed in that PR, it's just because it's high level).

This also relates to your question below about UUID. What I would suggest is to add a section, "Going to production", where you highlight some things:

* "Note on Python multi-threading" could be moved here.
* The note about error handling too
* Selecting a UUID. Basically a new random one which stays constant for that application.
* How `priority` interacts with a server based Update Control Map, or one from a different process.
* What happens on map expiry, how `on_map_expire` works, and why refreshing is necessary.

oleorhagen replied:
Could we not have this as the base-map with priority `0`, or similar, and then the other maps could have `Priority=1`, and then we can remove all but the state they are overriding?
So:
```    
    UPDATE_CONTROL_MAP_CONTINUE_INSTALL = """{
    	"priority": 1,
    "states": {   
    		"ArtifactInstall_Enter": {
    			"action": "continue"
    		},
    	},
    	"id": "01234567-89ab-cdef-0123-456789abcdef"
}
```

lluiscampos replied:
@kacf Okey about all that points, I added them to a new section.

@oleorhagen I think makes the example code a bit more advance. I have however added a not in the new section on how to use priorities and that the user might want to set different priorities on different maps.

lluiscampos commented on examples/update_flow_control/update_flow_control.py:
```diff
@@ -0,0 +1,254 @@
+			"action": "pause"
+		}
+	},
+	"id": "01234567-89ab-cdef-0123-456789abcdef"
```
Should we use a random `UUID`?

kacf replied:
I think this is fine. See also comment about `priority`.

oleorhagen replied:
Hmm, from the example perspective I would not consider it necessary, but if we are to promote proper usage, maybe we should? 
(sry, answering a question with a question :p)

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+logo](https://raw.githubusercontent.com/mendersoftware/mender/master/mender_logo.png)
+
+This repository contains simple example applications in Python that communicate
+with the Mender client on a device via DBus to demonstrate some of the use
```
Nit:
`D-Bus`

lluiscampos replied:
I started with `D-Bus` but then realized that the library we are using, `PyDBus`, used `DBus` on its documentation. Anyway, to be consistent with our own docs from docs.mender.io I'll change it to `D-Bus`.

(same for everywhere else you have pointed this out!)

Edit: I meant `D-Bus`!

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+The code in this repository is meant to educate and showcase the Mender DBus
+APIs capabilities. Please keep in mind that for simplicity error handling and
+failure recovery is not implemented in most cases, and we strongly discourage to
```
Suggestion:
`discourage the use of`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+## Documentation
+
+### APIs documentation
```
Suggestion:
`API documentation`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+### APIs documentation
+
+The DBus APIs of the Mender client are documented in the [Device-side API
```
`D-Bus`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+### Python PyDBus package
+
+The examples in this repository are using PyDBus package, which provides a
```
Nit: `the PyDBus`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+### Python PyDBus package
+
+The examples in this repository are using PyDBus package, which provides a
+modern and pythonic abstraction of the DBus API. Browse its [documentation
```
Suggestion:
`You can browse`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+here](https://pydbus.readthedocs.io/en/latest/index.html) and the [source code
+here](https://github.com/LEW21/pydbus).
+
+If you are new to DBus, you might find helpful this article from PyDBus
```
Suggestion:
`you might find [this article](...) from the PyDBus documentation helpful.`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+documentation: [A Pythonic way to understand Dbus object paths, interface names,
+etc.](https://pydbus.readthedocs.io/en/latest/dbusaddressing.html)
+
+## Installation
```
Suggestion:
`## Install`

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+### Requirements
+
+The examples require Python 3, package manager `pip`, and two development
```
Suggestion:
`, it's package manager`

oleorhagen commented on README.md:
````diff
@@ -1 +1,106 @@
+apt-get install -y python3 python3-pip libcairo2-dev libgirepository1.0-dev
+```
+
+TODO: check if they are installed by default in Raspberry Pi OS
````
Forget me not :)

lluiscampos replied:
:smile:

oleorhagen commented on README.md:
```diff
@@ -1 +1,106 @@
+
+TODO: check if they are installed by default in Raspberry Pi OS
+
+### Instructions
```
Usage?

lluiscampos replied:
Yes, better. Thanks

lluiscampos replied:
Well... not really. These are the actual install instructions (before was just requirements). Maybe `Install` here (even if it is repeating the one level up header, I think it makes sense).

I'll try that

oleorhagen commented on README.md:
````diff
@@ -1 +1,106 @@
+mender-update-flow-control
+```
+
+## Contributing
````
Suggestion, just link to the contributions file in the mender repo, maybe? :thinking:

lluiscampos replied:
I am leaving it (but updating the IRC bit) for the sake of consistency.

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,109 @@
+"pause" on every possible state. These are `ArtifactInstall_Enter`,
+`ArtifactReboot_Enter`, and `ArtifactCommit_Enter`. Learn more about the Mender
+client states in the [States section from Mender Docs
+page](https://docs.mender.io/overview/state-script#states).
```
Suggestion:
Remove the page, so just:
`[states section](...) from Mender Docs.`

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,109 @@
+page](https://docs.mender.io/overview/state-script#states).
+
+Then the application waits for user input. The user can select a new Update
+Control Map to send to the Mender client. Each of the possible maps set a one of
```
Suggestion:
Each of the control maps can override the previously set maps, and in this way force continue any ongoing update.

oleorhagen commented on examples/update_flow_control/README.md:
````diff
@@ -0,0 +1,109 @@
+mender-update-flow-control
+```
+
+Which initializes the application, setting the "pause" all Update Control Map,
````
Suggestion:
`sets all the Update Control Maps to "pause"`

lluiscampos replied:
I went with `sets all the states in the Update Control Map to "pause"`

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,109 @@
+
+## Implementation
+
+The example application implements to threads that communicate with a queue, one
```
`s/to/two`

oleorhagen reviewed (commented):
I will finish off  the rest later.

Mostly only found time for Readme this far :)

lluiscampos commented:
> I will finish off the rest later.
> 
> Mostly only found time for Readme this far :)

Thanks anyway. I have something to start with!

Yesterday I learned that we had a separate task for the README, so I'll review the acceptance criteria in that other task and make sure what I wrote here covers it. So I might need to rework them a bit anyway.

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,109 @@
+## Implementation
+
+The example application implements to threads that communicate with a queue, one
+thread responsible to manage the Mender DBus API communication, an another
```
`s/to manage/for managing/`

oleorhagen commented:
> Yesterday I learned that we had a separate task for the README, so I'll review the acceptance criteria in that other task and make sure what I wrote here covers it. So I might need to rework them a bit anyway.

Multitasking! I like it!

oleorhagen commented on .gitlab-ci.yml:
```diff
@@ -1,6 +1,25 @@
+    - pip install isort
+    - pip install mypy
+    - pip install pylint
+    - pip install .
```
:thinking: 
Maybe install the requirements before the app itself?

lluiscampos replied:
Well, I want to test that `setup.py` has the required requirements there, so that installing the app install also the requirements. Maybe I should remove `requirements.txt` file, as it is unused (I don't think we will add dependabot to this repo, and if so we should modify `setup.py` to read the reqs from the txt)

lluiscampos replied:
I'll remove the requirements.txt

oleorhagen commented on .gitlab-ci.yml:
```diff
@@ -1,6 +1,25 @@
+    - pip install .
+    - pip install -r requirements.txt
+  script:
+    - isort --check-only --line-length 88 --wrap-length 88 examples
```
Love it!
:medal_military: 
We've as a company have become a lot better at using these sorts of tools lately I feel :)

lluiscampos replied:
Well... you introduced these! I am only following your path :running_man:

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,108 @@
+## Implementation
+
+The example application implements two threads that communicate with two queues;
+one thread responsible to manage the Mender DBus API communication, an another
```
`one thread _is_ responsible`

oleorhagen replied:
`for managing`

oleorhagen replied:
`and another is responsible for dealing [...]`

lluiscampos replied:
Sorry :weary: Thanks for your patience on these typos

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,108 @@
+  the map needs refreshing. On DBus errors, the error message is reported in the
+  corresponding queue and the thread exists.
+
+* `do_main_interactive`: print the selection menu to the user and ask for action
```
`for what action [...]`

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,109 @@
+-----------------------
+What do you want to do?
+  0) Pause on all states
+  1) Continue with Installing new software
```
I'm not sure where it fits best, but I would include a mention about `on_action_executed`. What it does is to *atomically* execute the action of the map, and then set the next action, so that you don't risk race conditions when continuing an update and you want to pause the next update.

For example, with:
```
{
    "states": {
        "ArtifactInstall_Enter": {
            "action": "continue",
            "on_action_executed": "pause"
        },
    },
    "id": "01234567-89ab-cdef-0123-456789abcdef"
}
```

you don't need to set a new map to reinstate a pause. This should be our official recommendation.

lluiscampos replied:
:+1: 

I'll add this in the new section "Going to production" in the documentation, not in the code.

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,108 @@
+* `do_main_interactive`: print the selection menu to the user and ask for action
+  to do next. When the user selects an action, the corresponding map is sent to
+  the que for the DBus control thread to handle. It constantly refreshes the
+  terminal to always display the current map on top and the user options menu.
```
Q:
`on top of _the_` ? :thinking:

lluiscampos replied:
Yes

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,108 @@
+
+Python (or more specifically, the common implementation CPython) does not
+support concurrent multi-threading. Read more information in [threading module
+documentation](https://docs.python.org/3/library/threading.html) and [Python
```
`and the`

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,108 @@
+* `dbus-daemon` is running in your system
+* Mender client is 2.7.0 or newer
+* `DBus` APIs are enabled in the Mender client (they are enabled by default)
+* The `io.mender.UpdateManager.conf` policy file is installed in your system
```
And has the right privileges set, no?

lluiscampos replied:
I will write exactly that, "and has the right privileged", but I honestly don't know which are these.

oleorhagen replied:
It's which users `D-Bus` grants access to the bus

oleorhagen commented on examples/update_flow_control/update_flow_control.py:
```diff
@@ -0,0 +1,254 @@
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+"""example application to control the Update flow of the Mender client
```
```suggestion
#
# example application to control the Update flow of the Mender client
```

lluiscampos replied:
Mmmm. But what I am trying to achieve with this is to have a module level docstring (pylint will complain if I don't)

oleorhagen replied:
Oh, I see :)
All good then

oleorhagen reviewed (commented):
Real nice job, and this is just an example app!

I think it looks real good.

Left with two main chains of thought:

1. Ur writing is a little light on articles
2. Is it not written `D-Bus`?  ref: https://www.freedesktop.org/wiki/Software/dbus/

lluiscampos commented on examples/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 Northern.tech AS
```
Good catch. I wonder why was not caught by the CI license check...

oleorhagen replied:
:scream: :bug: :beetle:

lluiscampos commented:
New rework. Applied most of your feedback.


> Originally I was not planning to do any kind of error handling (first commit) but then decided to add some basic handling: to print the error and exit (second commit). Do you guys think the code gets "too complicated" with two queues?
> 
> If so, I will drop the second commit. If not, I will squash into a single commit.

No one has commented on this; hence I understand the code is not complicated an I have squashed it in :+1: 

@kacf ,

> What I would suggest is to add a section, "Going to production", where you highlight some things:

Added this new section. This is the main change of the rework, really. Please review in full this part.


@oleorhagen ,

> Multitasking! I like it!

Rather not reading all the tasks in the epic before starting a task :sweat_smile: 

> Real nice job, and this is just an example app!
> 
> I think it looks real good.
> 

Thanks :blush: 

> Left with two main chains of thought:
> 
>     1. Ur writing is a little light on articles

Thanks for all the nitpicks and for this comment. I have to improve my writing :man_technologist: 
> 
>     2. Is it not written `D-Bus`?  ref: https://www.freedesktop.org/wiki/Software/dbus/

Yes, I changed it everywhere.

lluiscampos commented:
> Forget me not :)

This is not forgotten. I will play with the example tomorrow on my RPi 4 :pie:

kacf commented on README.md:
```diff
@@ -1 +1,105 @@
+the code in this repository. You can find an introduction [introduction to D-Bus
+here](https://www.freedesktop.org/wiki/IntroductionToDBus/).
+
+### Python PyD-Bus package
```
Maybe one search-n-replace too much? We should not replace official names, should we? :-)

lluiscampos replied:
Yes, indeed. Thanks!

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+## Going to production
+
+This example is meant to educate and show how the Mender D-Bus API works. There
+are few aspects that you need to take into consideration when writing a production
```
`s/few/a few/`

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+
+This example uses a dummy [Universally Unique Identifier, or
+UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier). Make sure to
+chose a new random UUID for your application that will stay constant.
```
`s/chose/choose/`

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+
+### Error handling and failure recovery
+
+The code in this example just prints the message and exists on any D-Bus error.
```
`s/exists/exits/`

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+and by default the Mender client will continue on any step (assuming there no
+other maps). The application can however customize what happens when a map
+expires using its `on_map_expire` attribute in the map, which can be set to
+`continue`, `force_continue`, or `fail`.
```
Should we write that the motivation for expiring maps is to have a last resort when a frontend process has crashed or is hanging? Not sure if it makes it clearer, or if I'm just too close to the design and want to tell everything. I leave it to you to decide, since you have a lower bias.

lluiscampos replied:
It might be too much of a detail, but I'll add a sentence. Makes sense to my "lower bias" point of view.

lluiscampos replied:
I rewrote this paragraph like:

```
When a map expires it is not taken into account anymore for a Mender update and
by default the Mender client will continue on any step (assuming there no other
maps). This is a safety mechanism for the device to always be able to be updated
when a frontend process has crashed or is hanging.
```

kacf replied:
:+1:

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+# Update Control Map priority
+
+This example uses priority `-1` in all the maps being set. You may want to use
+different priority, or a combination of multiple priorities for different maps,
```
`s/different/a different/`

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+If multiple maps exist for the same priority, the order of precedence is `fail`,
+`pause`, `force_continue`, else priority takes precedence. `continue` is only
+executed if no other action exists at any priority level. The Mender client then
+applies the resulting Update Control Map.
```
You can also write:

```
Maps that are set using the Mender Web UI usually have priority `0`,
unless the deployment endpoint has been called manually to set a
different priority.
```

kacf commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,178 @@
+
+# Update Control Map one-time action
+
+Another interesting field in the map that might be interesting to know about is
```
Double "interesting". I would remove one of them.

kacf reviewed (approved):
Looks great! I have only nitpicks/suggestions and will pre-approve!

oleorhagen commented on README.md:
```diff
@@ -1 +1,105 @@
+### D-Bus documentation
+
+A basic understanding of the D-Bus concepts are required in order to understand
+the code in this repository. You can find an introduction [introduction to D-Bus
```
introduction introduction

oleorhagen commented on examples/update_flow_control/README.md:
```diff
@@ -0,0 +1,184 @@
+
+When multiple Update Control Maps are active, the Mender client will merge them.
+If multiple maps exist for the same priority, the order of precedence is `fail`,
+`pause`, `force_continue`, else priority takes precedence. `continue` is only
```
priority always takes precedence, but for maps with the same priority, it is as you describe here

lluiscampos replied:
Thanks, I'll rephrase.

oleorhagen commented on examples/update_flow_control/update_flow_control.py:
```diff
@@ -0,0 +1,254 @@
+    next_refresh_at = datetime.now() + timedelta(0, refresh_timeout - 1)
+
+    err_msg = ""
+    while err_msg == "":
```
Pythonistas would write:
`while not err_msg`

oleorhagen reviewed (approved):
The readme is looking so good, we should think about re-using parts here for our docs IMO

lluiscampos commented:
> > Forget me not :)
> 
> This is not forgotten. I will play with the example tomorrow on my RPi 4 pie

It works nicely on a RPi4. I added one more possible error in `Troubleshooting` section and slightly adjusted the requirements (adding git and removing the dev libraries, not needed after all).

Good to go! :rocket:
